[Merged by Bors] - feat(Algebra/Category/Ring): Tensor product over Z is coproduct in `CommRingCat`

justus-springer opened the pull request:
Show that the tensor product over the integers is a coproduct in the category of commutative rings. This is needed for constructing a `ChosenFiniteProducts` instance for the category of affine schemes. It would be nice to have the more general statement for commutative `R`-algebras, but it seems like mathlib does not define the category of commutative `R`-algebras yet.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

jcommelin commented on Mathlib/Algebra/Category/Ring/BinaryCoproducts.lean:
```diff
@@ -0,0 +1,91 @@
+    · simp only [forget_obj, pair_obj_left, tensorProductCocone_pt, Functor.const_obj_obj,
+        BinaryCofan.ι_app_left, pair_obj_right, AlgHom.toRingHom_eq_coe, CommRingCat.comp_apply,
+        coe_of]
+      erw [Algebra.TensorProduct.lift_tmul (hfg := fun _ _ => by apply Commute.all)]
```
Do you understand why this `erw` is needed? Is there a missing lemma?
Or maybe it goes away after merging master? Because there were refactors or concrete categories like `CommRingCat` in the past days.

justus-springer replied:
No I don't know exactly why it's needed. I'll merge master and look at it again.

justus-springer replied:
You're right, simp closes the goal now (it did not originally as far as I recall).

justus-springer commented:
As @erdOne suggested in #19977, I moved the content to the file `Constructions.lean`, which already contains similar constructions of colimits in `CommRingCat`. I also changed some names to better match the existing content.

jcommelin commented on Mathlib/Algebra/Category/Ring/Constructions.lean:
```diff
@@ -116,6 +117,60 @@ lemma isPushout_tensorProduct (R A B : Type u) [CommRing R] [CommRing A] [CommRi
+    apply Subtype.ext
+    rw [Algebra.TensorProduct.liftEquiv_symm_apply_coe, Prod.mk.injEq]
+    constructor
+    · ext a
+      dsimp
+      rw [map_one, mul_one]
+      have : _ = s.inl := hm (Discrete.mk WalkingPair.left)
+      rw [←this]
+      rfl
+    · ext b
+      dsimp
+      rw [map_one, one_mul]
+      have : _ = s.inr := hm (Discrete.mk WalkingPair.right)
+      rw [←this]
+      rfl
```
```suggestion
    · ext a
      simp [map_one, mul_one, ←hm (Discrete.mk WalkingPair.left)]
    · ext b
      simp [map_one, mul_one, ←hm (Discrete.mk WalkingPair.right)]
```

jcommelin reviewed (commented):
bors d+

jcommelin commented:
Thanks :tada:

bors merge
